python/ray/data/dataset.py

```diff
@@ -613,6 +613,55 @@ def do_shuffle(block_list, clear_input_blocks: bool, block_udf, remote_args):
         )
         return Dataset(plan, self._epoch, self._lazy)
 
+    def random_sample(
+        self, fraction: float, *, seed: Optional[int] = None
+    ) -> "Dataset[T]":
+        """Randomly samples a fraction of the elements of this dataset.
+
+        Note that the exact number of elements returned is not guaranteed.
+
+        Examples:
+            >>> import ray
+            >>> ds = ray.data.range(100) # doctest: +SKIP
+            >>> ds.random_sample(0.1) # doctest: +SKIP
+            >>> ds.random_sample(0.2, seed=12345) # doctest: +SKIP
+
+        Args:
+            fraction: The fraction of elements to sample.
+            seed: Seeds the python random pRNG generator.
+
+        Returns:
+            Returns a Dataset containing the sampled elements.
+        """
+        import random
+        import math
+        import pyarrow as pa
+        import pandas as pd
+
+        if self.num_blocks() == 0:
+            raise ValueError("Cannot sample from an empty dataset.")
+
+        if fraction < 0 or fraction > 1:
+            raise ValueError("Fraction must be between 0 and 1.")
+
+        if seed:
+            random.seed(seed)
+
+        def process_batch(batch):
+            if isinstance(batch, list):
+                return random.sample(batch, math.ceil(len(batch) * fraction))
```

Comment:
This seems not address the num_rows_in_block * x% < 1 case correctly, for example, if there are 5 blocks where each has 1 number, and we sample 20% from this dataset -- this will return all 5 numbers.

I think an algorithm that may work better is like this, for example, there is just 1 row and fraction=0.2, we return empty result with 80% probability and return that sing row as result with 20% probability; then the expected number of rows returned is 0.2.

Reply:
Yup, I committed that algorithm previously. I'll just revert back to it.

Reply:
46290c1a3ce1a9762254e9f9e7c9904f2a757bd0